### PR TITLE
Gather confusion matrix across GPUs in DDP validation

### LINKED ISSUE #24796: 在多卡上训练混淆矩阵没有合并
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

gather_stats(self) 这个函数只在终端输出验证集数量时合并了不同卡的验证集，但在画混淆矩阵时没有将不同卡上的数据合并，只输出0卡上的数据，问题代码ultralytics/models/yolo/detect/val.py

### Environment

[ultralytics-disk-cache-hook] enabled, ultralytics=8.4.0, cache root: /tmp/ultralytics-disk-cache, image_disk_cache=True, dataset_meta_cache=True
New https://pypi.org/project/ultralytics/8.4.63 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.4.0 🚀 Python-3.10.19 torch-2.10.0.dev20251208+cu128 CUDA:11 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:10 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:9 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:8 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:7 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:6 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:5 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:4 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:3 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:2 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:1 (NVIDIA GeForce RTX 5090, 32109MiB)
                                                                  CUDA:0 (NVIDIA GeForce RTX 5090, 32109MiB)

### Minimal Reproducible Example

ultralytics/models/yolo/detect/val.py:222

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Y-T-G**
Closes #24796 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves distributed detection validation by correctly combining confusion matrix results across multiple GPUs/devices 📈🖥️

### 📊 Key Changes
- Added synchronization logic in `ultralytics/models/yolo/detect/val.py` for the confusion matrix during distributed validation.
- When `plots=True` and distributed training/validation is active, the confusion matrix is:
  - converted to a PyTorch tensor on the active device,
  - reduced across all processes with `dist.reduce(...)`,
  - and restored on the main process (`RANK == 0`) as a NumPy array.
- This change only affects confusion matrix plotting/aggregation in multi-process environments, not the core metric computation flow.

### 🎯 Purpose & Impact
- Ensures confusion matrix plots reflect results from **all distributed workers**, not just a single process ✅
- Makes validation outputs more accurate and trustworthy when using multi-GPU or distributed setups 🚀
- Helps users better analyze model errors during large-scale training and validation runs 📊
- Has minimal impact on single-GPU or non-distributed workflows, where behavior remains effectively unchanged 👍

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with the review process—an engineer will assist you soon. Please review the checklist below to help ensure a smooth integration of your DDP validation fix. 🛠️📊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change is well-scoped and the aggregation approach makes sense, but there is one blocking regression: the new confusion-matrix gather path is not gated to DDP, so non-distributed validation will crash when plots are enabled.

💬 Posted 1 inline comment

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Clean, well-scoped fix. The DDP confusion-matrix gather is limited to plotting runs, uses the existing rank-0 aggregation pattern, and the added logic looks correct for producing full-run validation visualizations without affecting non-plot workflows.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged — thank you @Y-T-G, with contributions from @onuralpszr and @glenn-jocher!

“As our case is new, so we must think anew.” — Abraham Lincoln

This update is a great example of that mindset: improving distributed validation so confusion matrix plots reflect results from all workers makes multi-GPU analysis more accurate, reliable, and useful for everyone training at scale. A clean, focused fix with real impact.

We appreciate the thoughtful contribution and collaboration here 🚀
